File: experiments/archive_me_legacy_flutter/apps/voicememory_mobile/lib/features/archive_synthesis/archive_synthesis_hash.dart
import 'dart:convert';

import 'package:crypto/crypto.dart';

/// Stable archive hash for cache keys (matches server canonical JSON).
String computeArchiveHashFromPack(Map<String, dynamic> pack) {
  final sorted = _sortKeys(pack);
  final canonical = jsonEncode(sorted);
  return sha256.convert(utf8.encode(canonical)).toString().substring(0, 32);
}

dynamic _sortKeys(dynamic value) {
  if (value == null || value is! Map && value is! List) return value;
  if (value is List) return value.map(_sortKeys).toList();
  final map = value as Map;
  final keys = map.keys.map((k) => k.toString()).toList()..sort();
  final out = <String, dynamic>{};
  for (final key in keys) {
    out[key] = _sortKeys(map[key]);
  }
  return out;
}

String synthesisCacheKey({
  required String userId,
  required String monthKey,
  required String archiveHash,
}) => '$userId|$monthKey|$archiveHash';

File: packages/shared/lib/archive-synthesis/archive-synthesis-hash.ts
import { createHash } from "crypto";

import type { ArchiveSynthesisPack } from "@/types/archive-synthesis";

/** Stable hash for cache keys — excludes client-supplied archiveHash if present. */
export function computeArchiveHashFromPack(pack: ArchiveSynthesisPack): string {
  const { ...body } = pack;
  const canonical = JSON.stringify(sortKeys(body));
  return createHash("sha256").update(canonical).digest("hex").slice(0, 32);
}

function sortKeys(value: unknown): unknown {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map(sortKeys);
  const obj = value as Record<string, unknown>;
  const sorted: Record<string, unknown> = {};
  for (const key of Object.keys(obj).sort()) {
    sorted[key] = sortKeys(obj[key]);
  }
  return sorted;
}

export function synthesisCacheKey(
  subject: string,
  monthKey: string,
  archiveHash: string,
): string {
  return `${subject}:${monthKey}:${archiveHash}`;
}
